pydantic/_internal/_generate_schema.py

```diff
@@ -533,36 +530,21 @@ def _apply_discriminator_to_union(
                 discriminator,
             )
         except _discriminated_union.MissingDefinitionForUnionRef:
-            # defer until defs are resolved
-            _discriminated_union.set_discriminator_in_metadata(
-                schema,
-                discriminator,
-            )
+            # defer until defs are resolved.
+            _discriminated_union.set_discriminator_in_metadata(schema, discriminator)
+            self.defs.used_deferred_discriminators = True
             return schema
 
-    class CollectedInvalid(Exception):
-        pass
-
-    def clean_schema(self, schema: CoreSchema) -> CoreSchema:
-        schema = self.collect_definitions(schema)
-        schema = simplify_schema_references(schema)
-        if collect_invalid_schemas(schema):
-            raise self.CollectedInvalid()
-        schema = _discriminated_union.apply_discriminators(schema)
+    def clean_schema(self, schema: CoreSchema, *, deep_copy: bool) -> CoreSchema:
+        """Cleans the schema contents by substituting definition-refs and discriminators. Given schema must be the
+        one just given by generate_schema. Does in place substituting to the schema by mutating it for performance
+        reasons. Use deep_copy to avoid mutating the original schema.
+        """
```

Comment:
Let's be even more explicit here.

Can we:
* Explain subsituting definition refs
* Explain why / how we add tagged info to discriminators?

This isn't your doing at all, but our internal `GenerateSchema` logic is poorly documented, so I think it's good to have thorough improvements as a standard for PRs here going forward.

Reply:
Done, added some basic examples